moved check allowed value parameters to package configuration

KonradUdoHannes opened the pull request.

KonradUdoHannes committed:
moved check allowed value parameters to package configuration

added tests and extended allowed value mapping

KonradUdoHannes committed:
fixes regarding the explicit loading of tidyverse functionality for pipeline script 1 and 2

KonradUdoHannes commented:
@pmayd the second commit in this PR already addresses a few fixes for explicit tidyverse calls, at least wrt. pipelline script 1 and 2. In case we want to keep this we should merge this PR, but even then ensuring script 3 works is still outstanding.

pmayd commented:
> @pmayd the second commit in this PR already addresses a few fixes for explicit tidyverse calls, at least wrt. pipelline script 1 and 2. In case we want to keep this we should merge this PR, but even then ensuring script 3 works is still outstanding.
> 

I will take care of this and do the work for the 3rd script, thanks Konrad!

pmayd commented:
@KonradUdoHannes I am not sure I understand, so you have added yourself now the package names like `dplyr::` and so on? Did you rebase/merge with the latest develop or how do you want to solve the merge conflicts? Ah but I see you marked this PR as still WIP

pmayd commented:
and I already corrected all the scripts before pushing to dev so I don't really understand why you had to do this..

pmayd commented:
Hm ok I see the things you changed are actually not on dev...so I don't understand how I was able to run the scripts then...I run tests and checks and everything

pmayd commented:
Well, remove the draft when you are ready and I will merge

pmayd commented:
ah parse_invalid_dates in no longer used!

pmayd commented on R/helper_patient_data_fix.R:
```diff
@@ -172,27 +172,89 @@ parse_dates <- function(date) {
-#' @param error_val value that is used to signal invalid values.
 #' @param id patient id.
+#' @param replace_invalid A flag controlling whether to also replace invalid .
+#' @param error_val value that is used to signal invalid values.
```
maybe better "value that is used to replace the invalid value"

pmayd commented on R/helper_patient_data_fix.R:
```diff
@@ -172,27 +172,89 @@ parse_dates <- function(date) {
+        if (replace_invalid) {
             logInfo("Replacing ", x, " with ", error_val, ".")
-            x <- error_val
+            return(error_val)
```
instead of two return statements at the end, we could just return x as last statement (or not at all if the if command is returning a value like in other programming languages, not sure about R)

KonradUdoHannes replied:
I'm not sure how to improve the overall control flow here. Right now I find the readability ok as the return statements are closely related to the if statements. My impression is that this would become less clear if I try to move return statements around. In the end the function is already a little bit too complex with three different return routes. I would leave it as it is we see a very concrete readability improvement.

pmayd replied:
Now it is more obvious as there is another function called

pmayd commented on R/helper_patient_data_fix.R:
```diff
@@ -172,27 +172,89 @@ parse_dates <- function(date) {
 #'
 #' @return Either x or "Other".
-check_allowed_values <- function(x, valid_values, error_val, id, col = "") {
+check_allowed_values <- function(x, valid_values, id, replace_invalid = TRUE, error_val = "Undefined", col = "") {
```
the naming is somewhat off. If the function is named `check_` something I would not expect a return value or a boolean, but we are returning either the original value or `x`. Maybe we should call it `validate_x` but I don't find a good name. Anyway, it is not the allowed values that we check but the invalid ones, so maybe at least we could rename it to `handle_invalid_values`.

KonradUdoHannes replied:
I was also wondering about this. And the fact that the function is very complex and does many things does not make it easier. In this version the function might actually modify valid values as well to align them with the provided validity list. For instance if `Foo` is a valid value `foo` will be considered valid and will be mapped to `Foo`. While this adds additional complexity I though it might add nicer downstream guarantees wrt. the "valid" values that are expected based on configuration files. Previously `foo` would have been valid and unchanged.

I was also thinking to maybe go for something as `validate_`  but havn't found a solution yet with which I'm happy.

pmayd commented on R/helper_patient_data_fix.R:
```diff
@@ -172,27 +172,89 @@ parse_dates <- function(date) {
+                !!column_name
+            )
+        )
+    } else {
```
this is a lot of code repetition. Maybe we can just use `if else` to set the value for the error_expr, and call the function only once because nothing else changes?

pmayd commented on reference_data/build_package_data.R:
```diff
@@ -0,0 +1,11 @@
+cleaning_config <- yaml::read_yaml(here::here("reference_data","data_cleaning.yaml"))
```
spaces after ","

KonradUdoHannes replied:
This is actually a symptom of a slightly larger issue that I would address seperately; #103

pmayd commented on reference_data/data_cleaning.yaml:
```diff
@@ -0,0 +1,65 @@
+status:
+  steps:
+    - type: basic_function
+      function_name: (stringr::str_to_lower)
```
just for my understanding, why do you have to put the function name in `()`?

pmayd reviewed (approved):
only minor comments/observations

KonradUdoHannes commented:
@pmayd Thanks for the review. I actually didn't mean for you to go through all the stuff while its still in draft since some details might have changed anyways leading to a bit of unneeded review effort. But since you where so nice to do it I'll address these points as well.

Regarding the explicit namespace referencing changes from the second commit. I'm not sure what the state of development is but this branch is already rebased on develop and is currently only ahead. When going through the namespace issues I also noticed that we seem to have quite a poor test coverage in this regard. At least the tests all past long before the scripts where running. It might also be the case that tests just handle namespaces differently, but I would hope that this is not the case.

KonradUdoHannes committed:
adjusted metaprogramming

KonradUdoHannes commented:
I addressed the discussed issues except for the `check_allowed_values` function name which I didn't find a better alternative for yet. Either way I removed it from WIP to a regular PR. This should allow us to also merge the namespace changes and other work can easily be done in additional PRs.

KonradUdoHannes commented:
@pmayd do you want to have another look, or shall we just merge?

pmayd commented:
@KonradUdoHannes I trust you and you have a very high quality and good understanding of design so I learn more from you than the other way around and yes reviews are also for things that one might have overlooked but honestly when it works it works it is nothing anyone will ever see again haha so it is totally fine to merge right away!
